application/src/main/java/org/thingsboard/server/service/cloud/KafkaUplinkMessageService.java

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright © 2016-2024 The Thingsboard Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.cloud;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.thingsboard.server.common.data.cloud.CloudEvent;
+import org.thingsboard.server.common.data.id.TenantId;
+import org.thingsboard.server.common.data.page.PageData;
+import org.thingsboard.server.dao.cloud.CloudEventService;
+
+@Slf4j
+public abstract class KafkaUplinkMessageService extends BaseUplinkMessageService {
+    @Autowired
+    protected CloudEventService cloudEventService;
+
+    public void processHandleMessages(TenantId tenantId) {
+        PageData<CloudEvent> cloudEvents = findCloudEvents(tenantId);
+
+        if (cloudEvents.getTotalElements() != 0) {
+            processCloudEvents(tenantId, cloudEvents);
+        }
+    }
+
+    public void processCloudEvents(TenantId tenantId, PageData<CloudEvent> newCloudEvents) {
+        PageData<CloudEvent> cloudEvents = null;
+
+        do {
+            cloudEvents = cloudEvents == null ? newCloudEvents : findCloudEvents(tenantId);
```

Comment:
`cloudEvents` is always null, isn't it ?))

Reply:
No, because cloudEvents is null only in intro the first iteration of the do-while loop. After that, it takes the value of newCloudEvents (the values that triggered this method), and in subsequent iterations, if there are more events, it takes the value of findCloudEvents (the new events that were found).

I think so:)

Reply:
No, because cloudEvents is null only in intro the first iteration of the do-while loop. After that, it takes the value of newCloudEvents (the values that triggered this method), and in subsequent iterations, if there are more events, it takes the value of findCloudEvents (the new events that were found).

I think so:)